### ISSUE TITLE
Add a filtering method other than completion styles/"forwarding" regular expressions, for "without-regexp"

### ISSUE THREAD

**Issue by daanturo** [minad/vertico#212]
Use case: apply filtering where it's impossible with regexp-based completion styles (`orderless` and built-in friends).
With `selectrum`, I was able to add "without-regexp" by setting `selectrum-refine-candidates-function`:
```lisp
;;;###autoload
(defun my-remove-matched-candidates (patterns candidates)
  (let ((re (my-group-or-regexps patterns)))
    ;; case-sensitive, to allow removals like [A-Z]
    (dlet ((case-fold-search nil))
      ;; incomplete regexps: just return the unmodified candidates
      (condition-case nil
          (-remove (lambda (cand)
                     (string-match-p re cand))
                   candidates)
        (invalid-regexp candidates)))))

;;;###autoload
(defun my-input-and-candidates-not-match-any-after-bang (input candidates)
  "If there are patterns after \"! \", return members of CANDIDATES who don't match any of those patterns,
and the modified input where \"! \"; else just return INPUT and
CANDIDATES."
  (let ((patterns (split-string input " " 'omit-nulls)))
    (if-let ((patterns-after! (cdr (member "!" patterns))))
        (list (substring input 0 (string-match-p "\\(\\`\\| \\)! " input))
              (my-remove-matched-candidates patterns-after! candidates))
      `(,input ,candidates))))

(defun my-selectrum-refine-candidates (input candidates)
  "Ignore candidates that match any regexps after \"!\"."
  (apply #'selectrum-refine-candidates-using-completions-styles
         (my-input-and-candidates-not-match-any-after-bang input candidates)))

(setq selectrum-refine-candidates-function #'my-selectrum-refine-candidates)
```
Then for example: typing 
`f1 f` `some  ! [0-9] [A-Z] -- buffers$ ^org-` 
will find functions that contains "some" in its name, excluding numbers, capital letters, internal symbols, ones end with "buffers", ones begin with "^org-" (this is quite important because `org` functions are irrelevant is this context). A single "!" here is better than typing "!" 5+ times, but that's not important.

For `vertico`, I have tried 

```lisp
(defun my-vertico-filter-a (&rest _)
  (cl-destructuring-bind (old-inp . old-pt) vertico--input
    (cl-destructuring-bind
        (new-inp new-cands)
        (my-input-and-candidates-not-match-any-after-bang old-inp vertico--candidates)
      (setq
       vertico--input (cons new-inp
                            ;; should we change the point here?
                            (min old-pt (length new-inp)))
       vertico--candidates (vertico-sort-history-length-alpha new-cands)))))

(advice-add #'vertico--update-candidates :after #'my-vertico-filter-a)
```
Unfortunately the above doesn't work, I'm not sure where else to put the advice.

In short, can you provide a place to set where filtering with "NOT /regexp/" can happen and if it's possible already, point it out?

This request may not be in `vertico`'s scope, but I find it substantially handy when needed; and it doesn't fit well into other complementary packages (outside of `orderless`'s capability, currently).

**Comment by minad** [comment]
Filtering is out if scope, it belongs to the completion styles. You can create your own completion style if needed.

**Comment by minad** [comment]
You can also try to post process by advising: https://github.com/minad/vertico/blob/a92b1e47ffe343e2c3096e2ea61af013a8a02af9/vertico.el#L268

However my recommendation would really be to go via completion styles and not put something on top.

**Comment by daanturo** [comment]
Thank to your suggestion I have noticed `completion-all-completions`'s STRING and PRED argument.
We can implement "without regexp" by defining completion styles while checking those (previously I thought it was impossible with only completion styles because `orderless` doesn't publicly allow tweaking PRED).

**Comment by minad** [comment]
@daanturo Yes, you could propose that here: https://github.com/oantolin/orderless/issues/88
